### PR TITLE
Authentication and exception handling bug fixes

### LINKED ISSUE #39: Re-auth not happening after token expires
Expiration isn't currently taken into account causing re-auth not to happen.

### PR DESCRIPTION
**Pull request by ginach**
- Fix re-auth for expired tokens
- Use code flow instead of token flow for WebAuthenticationBroker flow so refresh tokens will be returned
- Enable a way to set user's sign in name on the service info provider for SSO scenarios
- Set throw site in OneDriveException

Fixes #39, #44, #50, #55

### CODE REVIEW

File: src/OneDriveSdk.WinRT/Authentication/AdalAuthenticationProvider.cs [hunk 1]
```diff
@@ -60,10 +60,13 @@ public AdalAuthenticationProvider(ServiceInfo serviceInfo, AccountSession curren
         protected override async Task<IAuthenticationResult> AuthenticateResourceAsync(string resource)
         {
             IAuthenticationResult authenticationResult = null;
+            var userIdentifier = string.IsNullOrEmpty(this.serviceInfo.UserId)
```
Comment (peternied) [inline src/OneDriveSdk.WinRT/Authentication/AdalAuthenticationProvider.cs]: Refactor into a helper function

File: src/OneDriveSdk/Authentication/AuthenticationProvider.cs [hunk 2]
```diff
@@ -151,6 +150,66 @@ protected AccountSession GetAuthenticationResultFromCache()
...
+        {
+            if (this.ServiceInfo.WebAuthenticationUi != null)
+            {
+                returnUrl = returnUrl ?? this.ServiceInfo.ReturnUrl;
+
+                var requestUriStringBuilder = new StringBuilder();
+                requestUriStringBuilder.Append(this.ServiceInfo.AuthenticationServiceUrl);
+                requestUriStringBuilder.AppendFormat("?{0}={1}", Constants.Authentication.RedirectUriKeyName, returnUrl);
+                requestUriStringBuilder.AppendFormat("&{0}={1}", Constants.Authentication.ClientIdKeyName, this.ServiceInfo.AppId);
+                requestUriStringBuilder.AppendFormat("&{0}={1}", Constants.Authentication.ScopeKeyName, string.Join("%20", this.ServiceInfo.Scopes));
```
Comment (peternied) [inline src/OneDriveSdk/Authentication/AuthenticationProvider.cs]: Looks like you are using a url encoder in the next function, could we use that over here too, or unify how scopes are processed?

File: src/OneDriveSdk/Exceptions/Error.cs [hunk 3]
```diff
@@ -39,6 +39,9 @@ public class Error
         [DataMember(Name = "message", IsRequired = false, EmitDefaultValue = false)]
         public string Message { get; set; }
 
+        [DataMember(Name = "throwSite", IsRequired = false, EmitDefaultValue = false)]
+        public string ThrowSite { get; set; }
```
Comment (peternied) [inline src/OneDriveSdk/Exceptions/Error.cs]: We should add a ToString function to make sure this debugging throwsite is included in most views of the exception to help ferry it into issue reports

**Comment by peternied** [comment]
:+1:

File: src/OneDriveSdk/Exceptions/Error.cs (hunk 3 above)
Comment (ginach) [inline src/OneDriveSdk/Exceptions/Error.cs]: Done. Other feedback should be addressed now too.

**Pull request merged**